spinalcordtoolbox/deepseg/monai.py

```diff
@@ -0,0 +1,212 @@
+    network_class = mapping[segmentation_network_class_name]
+
+    conv_or_blocks_per_stage = {
+        'n_conv_per_stage'
+        if network_class != ResidualEncoderUNet else 'n_blocks_per_stage': plans["n_conv_per_stage_encoder"],
+        'n_conv_per_stage_decoder': plans["n_conv_per_stage_decoder"]
+    }
+
+    # network class name!!
+    model = network_class(
+        input_channels=num_input_channels,
+        n_stages=num_stages,
+        features_per_stage=[min(plans["UNet_base_num_features"] * 2 ** i,
+                                plans["unet_max_num_features"]) for i in range(num_stages)],
+        conv_op=conv_op,
+        kernel_sizes=plans["conv_kernel_sizes"],
+        strides=plans["pool_op_kernel_sizes"],
+        num_classes=num_classes,
+        deep_supervision=deep_supervision,
+        **conv_or_blocks_per_stage,
+        **kwargs[segmentation_network_class_name]
+    )
+    model.apply(InitWeights_He(1e-2))
+    if network_class == ResidualEncoderUNet:
+        model.apply(init_last_bn_before_add_to_0)
+
+    # this loop only takes about 0.2s on average on a CPU
+    chkp_path = os.path.join(path_model, "best_model_loss.ckpt")
+    checkpoint = torch.load(chkp_path, map_location=torch.device(device))["state_dict"]
+    # NOTE: remove the 'net.' prefix from the keys because of how the model was initialized in lightning
+    # https://discuss.pytorch.org/t/missing-keys-unexpected-keys-in-state-dict-when-loading-self-trained-model/22379/14
+    for key in list(checkpoint.keys()):
+        if 'net.' in key:
+            checkpoint[key.replace('net.', '')] = checkpoint[key]
+            del checkpoint[key]
+
+    # load the trained model weights
+    model.load_state_dict(checkpoint)
+    model.to(device)
+    model.eval()
+
+    return model
+
+
+class InitWeights_He(object):
+    def __init__(self, neg_slope=1e-2):
+        self.neg_slope = neg_slope
+
+    def __call__(self, module):
+        if isinstance(module, nn.Conv3d) or isinstance(module, nn.ConvTranspose3d):
+            module.weight = nn.init.kaiming_normal_(module.weight, a=self.neg_slope)
+            if module.bias is not None:
+                module.bias = nn.init.constant_(module.bias, 0)
+
+
+def prepare_data(path_image, path_out, crop_size=(64, 160, 320)):
+    # create a temporary datalist containing the image
+    # boiler plate keys to be defined in the MSD-style datalist
+    params = {
+        "description": "my-awesome-SC-image",
+        "labels": {"0": "background", "1": "soft-sc-seg"},
+        "modality": {"0": "MRI"},
+        "tensorImageSize": "3D",
+        "test": [{"image": path_image}]
+    }
+    dataset = os.path.join(path_out, "temp_msd_datalist.json")
+    with open(dataset, "w") as jsonFile:
+        jsonFile.write(json.dumps(params, indent=4, sort_keys=True))
+    test_files = load_decathlon_datalist(dataset, True, "test")
+    os.remove(dataset)
```

Comment:
I found it a little odd that we had to mimic a full dataset just to segment a single image using MONAI. 

I'm not sure if there's a simpler way to perform inference using a single image, but I thought I would point it out.

Reply:
Hmm, what do you mean? Like creating the whole `temp_msd_datalist.json` just to segment a single image? If yes, then yeah, it's definitely an overkill. I didn't have enough time then to look into this hence I went with what was already working. 

If I find a simpler way, how easy/difficult would it be to modify the code here?

Reply:
> Like creating the whole `temp_msd_datalist.json` just to segment a single image? If yes, then yeah, it's definitely an overkill.

Yes exactly! 

> If I find a simpler way, how easy/difficult would it be to modify the code here?

Very easy, I think! The MONAI code you wrote included this nicely isolated `prepare_data` function, so it should be easy to change just this function without affecting the rest of the PR. :)

Reply:
Update: I created a PR that contains the simplified inference script and tagged you as a reviewer. 
PR: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/100
Changes commit: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/100/commits/7ed4ad112810d20f683304ef45144d04ae244a92

Reply:
I have tested the changes in your PR by incorporating them into my SCT branch:

- `Dataset` changes: 1fc640249c5017eb42a330babee6bd08aadaa257
- clamping/binarization changes: 011d6f937bdfa8a40f97be341badcff6ad373b73 (note that the softseg test already specified `sct_deepseg` with the argument `-thr 0.5`, which matches the binarization you added in the above PR)

Everything seems to work well, with the softseg test having identical DICE score against our ground truth before/after the changes, so they LGTM. :)

Reply:
Nice, thank you, Joshua!